packages/core-react/src/context/reefKnotModalContext.tsx

```diff
@@ -0,0 +1,135 @@
+import React, {
+  ReactNode,
+  createContext,
+  useContext,
+  useMemo,
+  useState,
+} from 'react';
+import { useLocalStorage } from '../helpers/useLocalStorage';
+import { LS_KEY_TERMS_ACCEPTANCE } from '../constants';
+
+/// MODAL STATE
+
+type RequirementsData = {
+  icon?: ReactNode;
+  title?: string;
+  text?: ReactNode;
+};
+
+type ModalTypes =
+  | {
+      type: 'wallet' | 'ledger';
+      props?: undefined;
+    }
+  | {
+      type: 'eager';
+      props: {
+        tryConnection: () => Promise<any>;
+        initialError?: Error;
+      };
+    }
+  | {
+      type: 'requirements';
+      props: RequirementsData;
+    };
+
+type ModalResult = {
+  success: boolean;
+};
+
+type ModalStateEntry = {
+  onClose: (props: ModalResult) => void;
+} & ModalTypes;
+
+type ModalOpenParams = {
+  onClose?: (props: ModalResult) => void;
+} & ModalTypes;
+
+type AsyncModalOpenParams = Omit<ModalOpenParams, 'onClose'>;
+
+export type ReefKnotModalContextValue = {
+  openModal: (params: ModalOpenParams) => void;
+  openModalAsync: (params: AsyncModalOpenParams) => Promise<ModalResult>;
+  forceCloseAllModals: () => void;
+  closeModal: (result: ModalResult) => void;
+  modalStack: ModalStateEntry[];
+  currentModal: ModalStateEntry | undefined;
+  termsChecked: boolean;
+  setTermsChecked: (isChecked: boolean) => void;
+};
+
+const NOOP = () => {};
+
+const ReefKnotModalContext = createContext<ReefKnotModalContextValue | null>(
+  null,
+);
+ReefKnotModalContext.displayName = 'ReefKnotModalContext';
+
+export const useReefKnotModal = () => {
```

Comment:
Maybe it is better to move it in `hooks` dir as a separate file. At least, we have such logic for `ReefKnot` context.